rust/datafusion/src/logical_plan/dfschema.rs

```diff
@@ -0,0 +1,415 @@
+    pub fn field_with_qualified_name(
+        &self,
+        relation_name: &str,
+        name: &str,
+    ) -> Result<DFField> {
+        let matches: Vec<&DFField> = self
+            .fields
+            .iter()
+            .filter(|field| {
+                field.qualifier == Some(relation_name.to_string()) && field.name() == name
+            })
+            .collect();
+        match matches.len() {
+            0 => Err(DataFusionError::Plan(format!(
+                "No field named '{}.{}'",
+                relation_name, name
+            ))),
+            1 => Ok(matches[0].to_owned()),
+            _ => Err(DataFusionError::Plan(format!(
+                "Ambiguous reference to qualified field named '{}.{}'",
+                relation_name, name
+            ))),
+        }
+    }
+
+    /// Convert to an Arrow schema
+    pub fn to_arrow_schema(&self) -> SchemaRef {
+        SchemaRef::new(Schema::new(
+            self.fields.iter().map(|f| f.field.clone()).collect(),
+        ))
+    }
+}
+
+impl Display for DFSchema {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.fields
+                .iter()
+                .map(|field| field.qualified_name())
+                .collect::<Vec<String>>()
+                .join(", ")
+        )
+    }
+}
+
+/// DFField wraps an Arrow field and adds an optional qualifier
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct DFField {
+    /// Optional qualifier (usually a table or relation name)
+    qualifier: Option<String>,
+    /// Arrow field definition
+    field: Field,
+}
+
+impl DFField {
+    /// Creates a new `DFField`
+    pub fn new(
+        qualifier: Option<&str>,
+        name: &str,
+        data_type: DataType,
+        nullable: bool,
+    ) -> Self {
+        DFField {
+            qualifier: qualifier.map(|s| s.to_owned()),
+            field: Field::new(name, data_type, nullable),
+        }
+    }
+
+    /// Create an unqualified field from an existing Arrow field
+    pub fn from(field: Field) -> Self {
+        Self {
+            qualifier: None,
+            field,
+        }
+    }
+
+    /// Create a qualified field from an existing Arrow field
+    pub fn from_qualified(qualifier: &str, field: Field) -> Self {
+        Self {
+            qualifier: Some(qualifier.to_owned()),
+            field,
+        }
+    }
+
+    /// Returns an immutable reference to the `DFField`'s unqualified name
+    pub fn name(&self) -> &String {
+        &self.field.name()
+    }
+
+    /// Returns an immutable reference to the `DFField`'s data-type
+    pub fn data_type(&self) -> &DataType {
+        &self.field.data_type()
+    }
+
+    /// Indicates whether this `DFField` supports null values
+    pub fn is_nullable(&self) -> bool {
+        self.field.is_nullable()
+    }
+
+    /// Returns an immutable reference to the `DFField`'s qualified name
```

Comment:
As a nit, as written this returns a `String` not an immutable reference

Maybe we could use `Cow` in this case.

Something like this (untested):

```
    pub fn qualified_name(&self) -> Cow<'_, str> {
        if let Some(relation_name) = &self.qualifier {
            Cow::owned(format!("{}.{}", relation_name, self.field.name()))
        } else {
            Cow::Borrowed(self.field.name())
        }
    }
```

I doubt this makes any measurable difference performance wise, it it would be nice to clean up the comment to match the code

Reply:
Thanks. I have copy and pasted the comments from Arrow's Field and will update the comment here.